File: lua/robot-name.lua
local names = {}

local function rand()
    local a = string.char(math.random(97, 122))
    local b = string.char(math.random(97, 122))
    local c = math.random(100, 999)
    return string.format("%s%s%d", a, b, c)
end

local function gen()
    local name = rand()
    while names[name] ~= nil do
        name = rand()
    end
    table.insert(names, name)
    return name
end

local Robot = {
    new = function(_)
        return {
            name = gen(),
            reset = function(self)
                self.name = gen()
            end,
        }
    end,
}

return Robot

File: go/robot_name.go
package robotname

import (
	"errors"
	"math/rand"
)

// Define the Robot type here.
type Robot struct {
	name string
}

var names = map[string]bool{}

func genName() string {
	name := make([]rune, 5)
	name[0] = rune(rand.Intn(26)) + 'A'
	name[1] = rune(rand.Intn(26)) + 'A'
	name[2] = rune(rand.Intn(10)) + '0'
	name[3] = rune(rand.Intn(10)) + '0'
	name[4] = rune(rand.Intn(10)) + '0'
	return string(name)
}

const totalNames = 26 * 26 * 10 * 10 * 10

func (r *Robot) Name() (string, error) {
	if r.name != "" {
		return r.name, nil
	}

	if len(names) == totalNames {
		return "", errors.New("full")
	}

	r.name = genName()
	for names[r.name] {
		r.name = genName()
	}
	names[r.name] = true
	return r.name, nil
}

func (r *Robot) Reset() {
	delete(names, r.name)
	r.name = ""
}
